tiled/adapters/protocols.py

```diff
@@ -0,0 +1,144 @@
+import collections.abc
+import sys
+from abc import abstractmethod
+from typing import Any, Dict, List, Literal, Optional, Protocol, Tuple, Union
+
+import dask.dataframe
+import pandas
+import sparse
+from numpy.typing import NDArray
+
+from ..server.schemas import Principal
+from ..structures.array import ArrayStructure
+from ..structures.awkward import AwkwardStructure
+from ..structures.core import Spec, StructureFamily
+from ..structures.sparse import SparseStructure
+from ..structures.table import TableStructure
+from .awkward_directory_container import DirectoryContainer
+from .type_alliases import JSON, Filters, NDSlice, Scopes
+
+
+class BaseAdapter(Protocol):
+    @abstractmethod
+    def metadata(self) -> JSON:
+        pass
+
+    @abstractmethod
+    def specs(self) -> List[Spec]:
+        pass
+
+
+if sys.version_info < (3, 9):
```

Comment:
Is this PR the time to upgrade Tiled's minimum supported python version to SPEC0, which at present would be 3.10 or greater? We could then remove this Mapping stuff.

Reply:
One or two users out there have asked if we can give them a 3.8 (or even a 3.7) compatible release up through the first beta, so they have at least a basically usable client available in legacy environments.

I figure we'll fast-forward to 3.10 shortly after tagging beta.